flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableSink.java

```diff
@@ -0,0 +1,142 @@
+						String tableName, List<String> partitionCols) {
+		this.jobConf = jobConf;
+		this.rowTypeInfo = rowTypeInfo;
+		this.dbName = dbName;
+		this.tableName = tableName;
+		this.partitionCols = partitionCols;
+	}
+
+	@Override
+	public DataStreamSink<?> emitBoundedStream(DataStream<BaseRow> boundedStream, TableConfig tableConfig, ExecutionConfig executionConfig) {
+		// TODO: support partitioning
+		final boolean isPartitioned = false;
+		// TODO: support overwrite
+		final boolean overwrite = false;
+		HiveTablePartition hiveTablePartition;
+		HiveTableOutputFormat outputFormat;
+		IMetaStoreClient client = HMSClientFactory.create(new HiveConf(jobConf, HiveConf.class));
+		try {
+			Table table = client.getTable(dbName, tableName);
+			StorageDescriptor sd = table.getSd();
+			// here we use the sdLocation to store the output path of the job, which is always a staging dir
+			String sdLocation = sd.getLocation();
+			if (isPartitioned) {
+				// TODO: implement this
+			} else {
+				sd.setLocation(toStagingDir(sdLocation, jobConf));
+				hiveTablePartition = new HiveTablePartition(sd, null);
+			}
+			outputFormat = new HiveTableOutputFormat(jobConf, dbName, tableName, partitionCols,
+				rowTypeInfo, hiveTablePartition, MetaStoreUtils.getTableMetadata(table), overwrite);
+		} catch (TException e) {
+			throw new CatalogException("Failed to query Hive metastore", e);
+		} catch (IOException e) {
+			throw new RuntimeException("Failed to create staging dir", e);
+		} finally {
+			client.close();
+		}
+		return boundedStream.writeUsingOutputFormat(outputFormat);
+	}
+
+	@Override
+	public TypeInformation<BaseRow> getOutputType() {
+		InternalType[] internalTypes = new InternalType[rowTypeInfo.getArity()];
+		for (int i = 0; i < internalTypes.length; i++) {
+			internalTypes[i] = TypeConverters.createInternalTypeFromTypeInfo(rowTypeInfo.getTypeAt(i));
+		}
+		return new BaseRowTypeInfo(internalTypes, rowTypeInfo.getFieldNames());
+	}
+
+	@Override
+	public String[] getFieldNames() {
+		return rowTypeInfo.getFieldNames();
+	}
+
+	@Override
+	public TypeInformation<?>[] getFieldTypes() {
+		return rowTypeInfo.getFieldTypes();
+	}
+
+	@Override
+	public TableSink<BaseRow> configure(String[] fieldNames, TypeInformation<?>[] fieldTypes) {
+		return new HiveTableSink(jobConf, rowTypeInfo, dbName, tableName, partitionCols);
+	}
+
+	// get a staging dir associated with a final dir
+	private String toStagingDir(String finalDir, Configuration conf) throws IOException {
+		String res = finalDir;
+		if (!finalDir.endsWith(Path.SEPARATOR)) {
+			res += Path.SEPARATOR;
+		}
+		// TODO: may append something more meaningful than a timestamp, like query ID
+		res += ".staging_" + System.currentTimeMillis();
+		Path path = new Path(res);
+		FileSystem fs = path.getFileSystem(conf);
+		Preconditions.checkState(fs.exists(path) || fs.mkdirs(path), "Failed to create staging dir " + path);
```

Comment:
Not sure if checkState is a good use here. The method throws IOException, which seems more applicable than asserting a condition.

Reply:
It checks we're in a state that either the output dir exists, or we're able to create it. IOException also seems a reasonable choice. Will use it instead.